File: 02_proxy/javascript/Proxy.js
// SẾP: người quyết định mọi việc nhận request cuối
class Leader {
  approveOffer(offer) {
    if (offer > 10000) {
        console.log(`Reject: $${offer}`);
    } else {
        console.log(`Accept: $${offer}`);
    }
  }
}

// THƯ KÝ: người được uỷ quyền bởi SẾP
class Secretary {
  constructor() {
    this.leader = new Leader();
  }

  forwardRequest(offer) {
    // Khi THƯ KÝ nhận mọi request thì SẾP là người nhận được câu hỏi và giải quyết
    this.leader.approveOffer(offer);
  }
}

// DEVELOPER: người tạo ra yêu cầu offer gửi đến SẾP nhưng thông qua THƯ KÝ của SẾP
class Developer {
  constructor(offer) {
    this.offer = offer;
  }

  // DEVELOPER: gửi đến THƯ KÝ
  submitOffer(offer) {
    offer.forwardRequest(this.offer);
  }
}

const dev01 = new Developer(10001);
const dev02 = new Developer(5000);
dev01.submitOffer(new Secretary());
dev02.submitOffer(new Secretary());

File: 02_proxy/java/Proxy.java

public class Proxy {
    public static void main(String[] args) {
        Developer dev01 = new Developer(10001);
        Developer dev02 = new Developer(5000);

        dev01.submitOffer(new Secretary());
        dev02.submitOffer(new Secretary());
    }
}

class Leader {
    public void approveOffer(double offer) {
        if (offer > 10000) {
            System.out.println("Reject: $" + offer);
        } else {
            System.out.println("Accept: $" + offer);
        }
    }
}

class Secretary {
    private Leader leader;

    public Secretary() {
        this.leader = new Leader();
    }

    public void forwardRequest(double offer) {
        leader.approveOffer(offer);
    }
}

class Developer {
    private double offer;

    public Developer(double offer) {
        this.offer = offer;
    }

    public void submitOffer(Secretary secretary) {
        secretary.forwardRequest(offer);
    }
}
